### PR TITLE
#22845 list and filter experiments

### LINKED ISSUE #22845: [UI] Create List Experiments from a Page
List Experiments from a Page

Requirements:

- [X] List grouped by status all the experiments of a selected Page by `identifier` `GET /experiments/{key}`
- [X] Create a Filter by status
- [X] Filter set by default ALL
- [X] Create a no experiments component
- [X] Create a navbar with the Page title and back to Page Edit
- [X] Add a Create new Experiment (Disabled)
- [X] Change of status  Menu of experiments
- - [X] ENDED: Can be changed to ARCHIVE `PUT /experiments/{key}/_archive`
- - [X] SCHEDULED & DRAFT: Can be deleted `DELETE /experiments/{key}`. This step needs a confirmation pop up message.

Empty state
Screen Shot 2022-08-24 at 12 42 54 PM

List
Screen Shot 2022-08-24 at 12 44 24 PM

Change of Status (Icon not menu)
Screen Shot 2022-08-24 at 1 44 01 PM
- DRAFT/SCHEDULED can be deleted
- ENDED, can be archived
- Others no icon

### PR DESCRIPTION
**Pull request by oidacra**
List and filter experiments

### CODE REVIEW

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/components/dot-experiments-status-filter/dot-experiments-status-filter.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+@Pipe({ name: 'dm' })
+class MockDmPipe implements PipeTransform {
+    transform(value: string): string {
+        return value;
+    }
+}
+
+const selectOptions = ExperimentsStatusList;
+fdescribe('DotExperimentsStatusFilterComponent', () => {
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/components/dot-experiments-status-filter/dot-experiments-status-filter.component.spec.ts:20]: remove `f`

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/ui/experiments-header/dot-experiments-ui-header.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { DotExperimentsUiHeaderComponent } from './dot-experiments-ui-header.component';
+
+xdescribe('ExperimentsHeaderComponent', () => {
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/ui/experiments-header/dot-experiments-ui-header.component.spec.ts:5]: missing tests

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/components/dot-experiments-status-filter/dot-experiments-status-filter.component.spec.ts (hunk 1 above)
Comment (oidacra) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/components/dot-experiments-status-filter/dot-experiments-status-filter.component.spec.ts:20]: I will find a rule of lint to check that.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/dot-experiments-list.component.ts [hunk 3]
```diff
@@ -1,11 +1,52 @@
...
+
+    addNewExperiment() {
+        // To implement
+    }
+
+    archiveExperiment(experimentId: string) {
+        this.experimentsStore.archiveExperiment(experimentId);
+    }
+
+    deleteExperiment(experimentId: string) {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/dot-experiments-list.component.ts:49]: doc.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-shell/dot-experiments-shell.component.ts [hunk 4]
```diff
@@ -1,13 +1,47 @@
...
+import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
 import { DotExperimentsStore } from '../shared/stores/dot-experiments-store.service';
+import { ActivatedRoute, Router } from '@angular/router';
+import { DotExperimentsService } from '../shared/services/dot-experiments.service';
+import { Observable } from 'rxjs';
+import { LoadingState } from '@portlets/shared/models/shared-models';
+import { RenderedPageExperiments } from '@portlets/dot-experiments/shared/models/dot-experiments.model';
+import { DotPageRenderState } from '@portlets/dot-edit-page/shared/models';
+
+interface VmShellExperiments {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-shell/dot-experiments-shell.component.ts:10]: This should live in the store.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-shell/dot-experiments-shell.component.ts [hunk 5]
```diff
@@ -1,13 +1,47 @@
...
+    vmShell$: Observable<VmShellExperiments> = this.experimentsStore.vmShell$;
+
+    constructor(
+        private readonly experimentsStore: DotExperimentsStore,
+        private readonly route: ActivatedRoute,
+        private readonly router: Router
+    ) {}
+
+    ngOnInit() {
+        this.experimentsStore.initStore(this._getResolverExperimentsData(this.route));
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-shell/dot-experiments-shell.component.ts:32]: Same as before, you can have the `Router` at the store.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/models/dot-experiments.model.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    },
+    {
+        label: 'experimentspage.experiment.status.archived',
+        value: DotExperimentStatusList.ARCHIVED
+    },
+    {
+        label: 'experimentspage.experiment.status.scheduled',
+        value: DotExperimentStatusList.SCHEDULED
+    }
+];
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/models/dot-experiments.model.ts:51]: This is not a model per-se...

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        return this.http.get<DotCMSResponse<DotExperiment[]>>(URL).pipe(pluck('entity'));
+    }
+
+    archive(experimentId: string): Observable<DotExperiment[]> {
+        const URL = `/api/v1/experiments/${experimentId}/_archive`;
+
+        return this.http.put<DotCMSResponse<DotExperiment[]>>(URL, {}).pipe(pluck('entity'));
+    }
+
+    delete(experimentId: string): Observable<string | DotExperiment[]> {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts:24]: doc.

Comment (fmontes) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts:24]: Extract `/api/v1/experiments` as a const.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    get(pageId: string): Observable<DotExperiment[]> {
+        const URL = `/api/v1/experiments?pageId=${pageId}`;
+
+        return this.http.get<DotCMSResponse<DotExperiment[]>>(URL).pipe(pluck('entity'));
+    }
+
+    archive(experimentId: string): Observable<DotExperiment[]> {
+        const URL = `/api/v1/experiments/${experimentId}/_archive`;
+
+        return this.http.put<DotCMSResponse<DotExperiment[]>>(URL, {}).pipe(pluck('entity'));
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts:21]: OHHHHHHHHHHHH!!!! I like it.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts [hunk 9]
```diff
@@ -1,21 +1,175 @@
...
+    experimentsFilterStatus: Array<string>;
+    status: LoadingState;
 }
 
 const initialState: DotExperimentsState = {
     pageId: '',
-    state: LoadingState.INIT
+    pageTitle: '',
+    experiments: [],
+    experimentsFilterStatus: [
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts:27]:
```suggestion
    filterStatus: [
```

We are already in "experiment"

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts [hunk 10]
```diff
@@ -1,21 +1,175 @@
...
+
+@Injectable({ providedIn: 'root' })
 export class DotExperimentsStore extends ComponentStore<DotExperimentsState> {
-    constructor() {
+    // Selectors
+    readonly getStatus$ = this.select((state) => state.status);
+    readonly getSelectedPageTitle$ = this.select((state) => state.pageTitle);
+    readonly getSelectedPageId$ = this.select((state) => state.pageId);
+    readonly getExperiments$ = this.select((state) => state.experiments);
+    readonly getExperimentsFilterStatusList$ = this.select(
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts:51]: Don't need the `get` prefix here.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts [hunk 11]
```diff
@@ -1,21 +1,175 @@
...
+                .reduce<GroupedExperimentByStatus>((group, experiment) => {
+                    group[experiment.status] = group[experiment.status] ?? [];
+                    group[experiment.status].push(experiment);
+
+                    return group;
+                }, <GroupedExperimentByStatus>{})
+    );
+    readonly isLoading$: Observable<boolean> = this.select(
+        (state) => state.status === LoadingState.LOADING || state.status === LoadingState.INIT
+    );
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts:67]: kinda contradictory.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts [hunk 12]
```diff
@@ -1,21 +1,175 @@
...
+    readonly deleteExperimentById = this.updater((state, experimentId: string) => ({
+        ...state,
+        experiments: state.experiments.filter((exp) => exp.id != experimentId)
+    }));
+    readonly archiveExperimentById = this.updater((state, experimentId: string) => ({
+        ...state,
+        experiments: state.experiments.map((exp) =>
+            experimentId === exp.id ? { ...exp, status: DotExperimentStatusList.ARCHIVED } : exp
+        )
+    }));
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts:100]:
```suggestion
    //Updater
    readonly initStore = this.updater((state, renderedPage: RenderedPageExperiments) => ({
        ...state,
        pageId: renderedPage.identifier,
        pageTitle: renderedPage.title,
        status: LoadingState.INIT
    }));

    readonly setComponentStatus = this.updater((state, status: LoadingState) => ({
        ...state,
        status
    }));

    readonly setExperiments = this.updater((state, experiments: DotExperiment[]) => ({
        ...state,
        status: LoadingState.LOADED,
        experiments
    }));

    readonly setExperimentsFilterStatus = this.updater(
        (state, experimentsFilterStatus: Array<string>) => ({
            ...state,
            experimentsFilterStatus
        })
    );

    readonly deleteExperimentById = this.updater((state, experimentId: string) => ({
        ...state,
        experiments: state.experiments.filter((exp) => exp.id != experimentId)
    }));

    readonly archiveExperimentById = this.updater((state, experimentId: string) => ({
        ...state,
        experiments: state.experiments.map((exp) =>
            experimentId === exp.id ? { ...exp, status: DotExperimentStatusList.ARCHIVED } : exp
        )
    }));
```

We need to update this rule.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts [hunk 13]
```diff
@@ -1,21 +1,175 @@
...
+        this.getExperiments$,
+        this.getExperimentsFilteredAndGroupedByStatus$,
+        this.getExperimentsFilterStatusList$,
+        (isLoading, experiments, experimentsFiltered, experimentsFilterStatus) => ({
+            isLoading,
+            experiments,
+            experimentsFiltered,
+            experimentsFilterStatus
+        })
+    );
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/stores/dot-experiments-store.service.ts:170]: Looks like you need a store for the shell and another one or for the list, stores should not be shared between components.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/ui/experiments-header/dot-experiments-ui-header.component.scss [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    height: $dot-secondary-toolbar-height;
+    align-items: center;
+    background-color: $white;
+    border-bottom: 1px $gray-lighter;
+    color: #7e7b86;
+    font-size: $font-size-medium;
+    column-gap: 15px;
+    border-bottom: 1px solid $gray-lighter;
+    a {
+        padding-left: 15px;
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/ui/experiments-header/dot-experiments-ui-header.component.scss:14]: var

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/ui/experiments-header/dot-experiments-ui-header.component.scss [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+:host {
+    display: flex;
+    flex-direction: row;
+    height: $dot-secondary-toolbar-height;
+    align-items: center;
+    background-color: $white;
+    border-bottom: 1px $gray-lighter;
+    color: #7e7b86;
+    font-size: $font-size-medium;
+    column-gap: 15px;
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/ui/experiments-header/dot-experiments-ui-header.component.scss:11]: var

File: core-web/libs/dot-primeng-theme-styles/src/scss/_variables.scss [hunk 16]
```diff
@@ -54,6 +54,8 @@ $button-shadow-focus: 0 0 0 4px rgba(var(--color-sec_rgb), 0.2);
 // Tables
 $th-text-color: $gray-light;
 $table-border-color: $gray-light;
+$table-border-color-experiments: $gray-lighter;
+$th-text-color-experiments: $gray;
```
Comment (fmontes) [inline core-web/libs/dot-primeng-theme-styles/src/scss/_variables.scss:58]: can we have this in the experiment context? should we? is more a question.

File: core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss [hunk 17]
```diff
@@ -0,0 +1,91 @@
+@use "variables" as *;
+
+.p-confirm-popup {
+    position: absolute;
+    margin-top: 10px;
```
Comment (fmontes) [inline core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss:5]: var

File: core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    top: 0;
+    left: 0;
+
+    background: $white;
+    color: #495057;
```
Comment (fmontes) [inline core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss:10]: var

File: core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    .p-confirm-popup-message {
+        margin-left: $spacing-3;
+    }
+
+    .p-confirm-popup-footer {
+        padding: $spacing-2 $spacing-4;
+        display: flex;
+        column-gap: 5px;
```
Comment (fmontes) [inline core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss:26]: var

File: core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+        button {
+            margin: 0 0 $spacing-1 0;
+            width: auto;
+        }
+    }
+}
+
+.p-confirm-popup-flipped {
+    margin-top: 0;
+    margin-bottom: 10px;
```
Comment (fmontes) [inline core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss:38]: var

Comment (fmontes) [reply core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_confirmpopup.scss:38]: and so on.

**Review by fmontes — DISMISSED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/dot-experiments-list.component.ts [hunk 21]
```diff
@@ -1,11 +1,64 @@
...
+        private readonly dotMessagePipe: DotMessagePipe,
+        private readonly route: ActivatedRoute
+    ) {}
+
+    ngOnInit() {
+        const pageId = this.route.parent?.parent?.parent.snapshot.params.pageId;
+        this.dotExperimentsListStore.loadExperiments(pageId);
+    }
+
+    selectedStatusFilter($event: Array<string>) {
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/dot-experiments-list.component.ts:37]: doc?

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/dot-experiments-list.component.ts [hunk 22]
```diff
@@ -1,11 +1,64 @@
...
+     * Add experiment
+     * @param experiment
+     */
+    addNewExperiment() {
+        // To implement
+    }
+
+    /**
+     * Archive selected experiment
+     * @param experiment
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/dot-experiments-list.component.ts:51]: update doc

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/dot-experiments-list.component.ts [hunk 23]
```diff
@@ -1,11 +1,64 @@
...
+     */
+    archiveExperiment(experiment: DotExperiment) {
+        this.dotExperimentsListStore.archiveExperiment(experiment);
+    }
+
+    /**
+     * Delete selected experiment
+     * @param experiment
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/dot-experiments-list.component.ts:59]: update doc, add type

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+const API_ENDPOINT = '/api/v1/experiments';
+
+@Injectable()
+export class DotExperimentsService {
+    constructor(private readonly http: HttpClient) {}
+
+    /**
+     * Get an array of experiments of a pageId
+     * @param pageId
+     * @memberof DotExperimentsService
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts:17]: update doc

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+    get(pageId: string): Observable<DotExperiment[]> {
+        const URL = `${API_ENDPOINT}?pageId=${pageId}`;
+
+        return this.http.get<DotCMSResponse<DotExperiment[]>>(URL).pipe(pluck('entity'));
+    }
+
+    /**
+     * Archive an experiment with its experimentId
+     * @param experimentId
+     * @memberof DotExperimentsService
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts:28]: update doc

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    archive(experimentId: string): Observable<DotExperiment[]> {
+        const URL = `${API_ENDPOINT}/${experimentId}/_archive`;
+
+        return this.http.put<DotCMSResponse<DotExperiment[]>>(URL, {}).pipe(pluck('entity'));
+    }
+
+    /**
+     * Delete an experiment with its experimenId
+     * @param experimentId
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments.service.ts:38]: update doc

**Review by alfredo-dotcms — APPROVED** [review]

**Review by fmontes — CHANGES_REQUESTED** [review]

**Review by fmontes — APPROVED** [review]

**Pull request merged**